crates/loans/src/interest.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /// Update the exchange rate according to the totalCash, totalBorrows and totalSupply.
+    /// This function does not accrue interest before calculating the exchange rate.
+    /// exchangeRate = (totalCash + totalBorrows - totalReserves) / totalSupply
+    pub fn exchange_rate_stored(asset_id: AssetIdOf<T>) -> Result<Rate, DispatchError> {
+        let total_supply = Self::total_supply(asset_id)?;
+        let total_cash = Self::get_total_cash(asset_id);
+        let total_borrows = Self::total_borrows(asset_id);
+        let total_reserves = Self::total_reserves(asset_id);
+
+        Self::calculate_exchange_rate(total_supply, total_cash, total_borrows, total_reserves)
+    }
+
+    /// Calculate the borrowing utilization ratio of the specified market
+    ///
+    /// utilizationRatio = totalBorrows / (totalCash + totalBorrows − totalReserves)
+    pub(crate) fn calc_utilization_ratio(
+        cash: BalanceOf<T>,
+        borrows: BalanceOf<T>,
+        reserves: BalanceOf<T>,
+    ) -> Result<Ratio, DispatchError> {
+        // utilization ratio is 0 when there are no borrows
+        if borrows.is_zero() {
+            return Ok(Ratio::zero());
+        }
+        let total = cash
+            .checked_add(borrows)
+            .and_then(|r| r.checked_sub(reserves))
+            .ok_or(ArithmeticError::Overflow)?;
+
+        Ok(Ratio::from_rational(borrows, total))
+    }
+
+    /// The exchange rate should be greater than the `MinExchangeRate` storage value and less than
+    /// the `MaxExchangeRate` storage value.
+    /// This ensures the exchange rate cannot be attacked by a deposit so big that
+    /// subsequent deposits to receive zero lendTokens (because of rounding down). See this
+    /// PR for more details: https://github.com/parallel-finance/parallel/pull/1552/files
+    pub(crate) fn ensure_valid_exchange_rate(exchange_rate: Rate) -> DispatchResult {
+        ensure!(
+            exchange_rate >= Self::min_exchange_rate() && exchange_rate < Self::max_exchange_rate(),
+            Error::<T>::InvalidExchangeRate
+        );
+
+        Ok(())
+    }
+
+    pub(crate) fn update_last_accrued_interest_time(asset_id: AssetIdOf<T>, time: Timestamp) -> DispatchResult {
+        LastAccruedInterestTime::<T>::try_mutate(asset_id, |last_time| -> DispatchResult {
+            *last_time = time;
+            Ok(())
+        })
+    }
+
+    fn accrued_interest(borrow_rate: Rate, amount: BalanceOf<T>, delta_time: Timestamp) -> Option<BalanceOf<T>> {
+        borrow_rate
+            .checked_mul_int(amount)?
+            .checked_mul(delta_time.into())?
+            .checked_div(SECONDS_PER_YEAR.into())
+    }
+
+    fn increment_index(borrow_rate: Rate, index: Rate, delta_time: Timestamp) -> Option<Rate> {
+        borrow_rate
+            .checked_mul(&index)?
+            .checked_mul(&FixedU128::saturating_from_integer(delta_time))?
+            .checked_div(&FixedU128::saturating_from_integer(SECONDS_PER_YEAR))
+    }
```

Comment:
Why are we dividing by `SECONDS_PER_YEAR`? I can see these two functions are only called via RPC so I can only assume this is used to determine APY? Although I think there would be an implicit assumption that `Timestamp` is stored in seconds.

On a related note I can see that Substrate uses the "Julian" year, perhaps that is a standard we should consider adopting also elsewhere: https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/frame/staking/src/inflation.rs#L42

Reply:
This is how interest accrual was implemented by Parallel (the functions are used everywhere, not just in RPCs), but ultimately it doesn't matter that much. The interest rate model that decides interest rates based probably also assumes a fixed-length year. We should double-check but it might actually simplify things.

Reply:
Created a nice-to-have issue: https://github.com/interlay/interbtc/issues/758